src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -30,153 +29,171 @@ case class PortsMatch(hostPortsWithRole: Seq[Option[PortWithRole]]) {
 
-    (runSpec.portNumbers, portMappings) match {
-      case (Nil, None) => // optimization for empty special case
-        Some(Seq.empty)
+        case (ports, None) if requirePorts =>
+          findPortsInOffer(ports, failLog = true)
 
-      case (ports, Some(mappings)) =>
-        // We use the mappings from the containers if they are available and ignore any other port specification.
-        // We cannot warn about this because we autofill the ports field.
-        mappedPortRanges(mappings)
-
-      case (ports, None) if runSpec.requirePorts =>
-        findPortsInOffer(ports, failLog = true)
-
-      case (ports, None) =>
-        randomPorts(ports.size)
+        case (ports, None) =>
+          randomPorts(ports.size)
+      }
     }
-  }
 
-  /**
-    * Try to find supplied ports in offer. Returns `None` if not all ports were found.
-    */
-  private[this] def findPortsInOffer(requiredPorts: Seq[Int], failLog: Boolean): Option[Seq[Option[PortWithRole]]] = {
-    takeEnoughPortsOrNone(expectedSize = requiredPorts.size) {
-      requiredPorts.iterator.map { (port: Int) =>
-        offeredPortRanges.find(_.contains(port)).map { offeredRange =>
-          PortWithRole(offeredRange.role, port, offeredRange.reservation)
-        } orElse {
-          if (failLog)
-            log.info(
-              s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
-                s"Couldn't find host port $port (of ${requiredPorts.mkString(", ")}) " +
-                s"in any offered range for run spec [${runSpec.id}]")
-          None
+    /**
+      * Try to find supplied ports in offer. Returns `None` if not all ports were found.
+      */
+    def findPortsInOffer(requiredPorts: Seq[Int], failLog: Boolean): Option[Seq[Option[PortWithRole]]] = {
+      takeEnoughPortsOrNone(expectedSize = requiredPorts.size) {
+        requiredPorts.iterator.map { (port: Int) =>
+          offeredPortRanges.find(_.contains(port)).map { offeredRange =>
+            PortWithRole(offeredRange.role, port, offeredRange.reservation)
+          } orElse {
+            if (failLog)
+              log.info(
+                s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
+                  s"Couldn't find host port $port (of ${requiredPorts.mkString(", ")}) " +
+                  s"in any offered range for ${name}")
+            None
+          }
         }
       }
     }
-  }
 
-  /**
-    * Choose random ports from offer.
-    */
-  private[this] def randomPorts(numberOfPorts: Int): Option[Seq[Option[PortWithRole]]] = {
-    takeEnoughPortsOrNone(expectedSize = numberOfPorts) {
-      shuffledAvailablePorts.map(Some(_))
-    } orElse {
-      log.info(s"Offer [${offer.getId.getValue}]. $resourceSelector. " +
-        s"Couldn't find $numberOfPorts ports in offer for run spec [${runSpec.id}]")
-      None
+    /**
+      * Choose random ports from offer.
+      */
+    def randomPorts(numberOfPorts: Int): Option[Seq[Option[PortWithRole]]] = {
```

Comment:
Let's kick this improvement down the line. I'd like to focus on reworking all of the matchers into a common trait first, and then improve the internal implementation later.